Srikar Murali

CPTS 215 Fall 2017 

Professor Ludlow 

Micro Assignment #4 

10/26/2017


#1

You can make a queue using these function by using the push and pop function for the enqueue and dequeue methods. To do this you must first implement the queue as a list, and then when adding items to the queue just push them in at the front. When trying to dequeue items, just traverse the list with a loop until you get to the oldest item in the queue. You could use two stacks for this. To enqueue you should follow these steps. While stack1 is not empty, push everything from satck1 to stack2. Push x to stack1 (assuming size of stacks is unlimited). Push everything back to stack1. To dequeue follow these steps. If stack1 is empty then error. If not then pop an item from stack1 and return it.

#2

The worst case run time will be $O(n)$ when either using enqueue or dequeue depending on how you set up the queue.

#3

The pop operation will perform n dequeues for each of the enqueues. However there will probably be around $n^2$ pops.


In [92]:
##Queue List implementation
import random
import timeit
import time
import urllib.request

class Queue:
    def __init__(self):
        self.items = []
        self.dqCount = 0
    
    def isEmpty(self):
        return self.items == []
    
    def enqueue(self, item):
        self.items.insert(0, item)
    
    def dequeue(self):
        self.dqCount+=1
        return self.items.pop()
    
    def size(self):
        return len(self.items)
    
def main():
    n = 10000
    d = Queue()
    data = urllib.request.urlopen('https://raw.githubusercontent.com/gsprint23/cpts215/master/microassignments/files/words.txt').read()

    for i in range(len(data)):
        d.enqueue(data[i])

    start = time.time()
    
    while d.isEmpty() == False:
        d.dequeue()
    end = time.time()
    
    print(end - start)
    print('Dequeue Count', d.dqCount)

if __name__ == '__main__':
    main()

KeyboardInterrupt: 

#2/3 for list implementation of queue

For a list of size 100

Time

1. 0.0

2. 0.0

3. 0.0

4. 0.0

5. 0.0

Dequeue Count

1. 100

2. 100

3. 100

4. 100

5. 100

The list implementation is very fast of a list of size 100.

For a list of size 1000

Time

1. 0.001

2. 0.002

3. 0.0

4. 0.001

5. 0.00099

Dequeue Count

1. 1000

2. 1000

3. 1000

4. 1000

5. 1000

For a list of 1000 it is still very fast. The dequeue function has a runtime of $O(1)$ on average and $O(n)$ in the worst case.

For a list of size 10000

Time

1. 0.00613

2. 0.006504

3. 0.00701618

4. 0.006003

5. 0.006986

Dequeue Count

1. 10000

2. 10000

3. 10000

4. 10000

5. 10000

Even for a list of size 10000, the list version is still very fast.


When using the words.txt file, the runtime for the enqueue and dequeue functions took too long for this computer to handle. Thus, the list implementation is not as efficient as the linked list implementation, but is better than the two stacks implementation.

In [71]:
#Queue Implementation using two stacks
#Code for question 2 and 3

import timeit
import time
import urllib.request

class Stack():

    def __init__(self):
        self.stk = []

    def pop(self):
        """raises IndexError if you pop when it's empty"""
        return self.stk.pop()

    def push(self, elt):
        self.stk.append(elt)

    def is_empty(self):
        return len(self.stk) == 0

    def peek(self):
        if not self.stk.is_empty():
            return self.stk[-1]


class Queue():

    def __init__(self):
        self.q = Stack()  
        self.b = Stack()  
        self.front = None
        self.size = 0
        self.popCount = 0

    def is_empty(self):
        return self.q.is_empty()

    def peek(self):
        if self.q.is_empty():
            return None
        else:
            return self.front

    def enqueue(self, elt):
        self.front = elt
        self.q.push(elt)
        self.size+=1

    def dequeue(self):
        """raises IndexError if you dequeue from an empty queue"""
        while not self.q.is_empty() > 0:
            elt = self.q.pop()
            self.b.push(elt)
            self.popCount+=1
        val = self.b.pop()
        self.popCount+=1
        elt = None
        while not self.b.is_empty() > 0:
            elt = self.b.pop()
            self.popCount+=1
            self.q.push(elt)
        self.front = elt
        self.size-=1
        return val
    
    
def main():
    c = Queue()
    n = 100
    #for i in range(n):
     #   c.enqueue(i)
    start = time.time()
    data = urllib.request.urlopen('https://raw.githubusercontent.com/gsprint23/cpts215/master/microassignments/files/words.txt').read()


    for i in range(len(data)):
        c.enqueue(data[i])
    while c.is_empty() == False:
        c.dequeue()
    end = time.time()
    print(end - start)
    print("Pop count", c.popCount)
if __name__ == '__main__':
    main()

KeyboardInterrupt: 


#2/3 observations

The worst case runtime is when the dequeue function has to go through the whole queue to find the element to pop off. The worst case runtime is $O(n)$ and it is $O(n^2)$ for the pop function. 


#Results From Two Stacks Implentation of Queue

For a list of size 100

Times

1. 0.01099

2. 0.0110

3. 0.01251

4. 0.01301

5. 0.0135

Pop Counts

1. 10100

2. 10100

3. 10100

4. 10100

5. 10100

The times and pop counts seem to be fairy consistent with a list of size 100. For a list of size 1000

Times

1. 0.99195

2. 0.9878

3. 1.00982

4. 0.98901

5. 0.9717

Pop Count

1. 1001000

2. 1001000

3. 1001000

4. 1001000

5. 1001000

Again very similar.

When trying to use the dequeue function for a list of size 10000, it took too long and the computer couldn't handle it.

Also when trying to use the words.txt data to test the enqueue and dequeue functions, it took much too long for the computer to handle, thus showcasing that the runtime for the two stack implementation of a queue is much slower than the linked list approach.

In [70]:
##Linked list implementation of Stacks/Queues

import time
import urllib.request

class Node:
    def __init__(self, initdata):
        self.data = initdata
        self.next = None
        
    def getData(self):
        return self.data
    
    def getNext(self):
        return self.next
    
    def setData(self, newdata):
        self.data = newdata
        
    def setNext(self, newnext):
        self.next = newnext
        
class Queue:
    def __init__(self):
        self.front = None
        self.back = None
        self.numlinks = 0
        self.dqCount = 0
        
    def enqueue(self, data):
        newNode = Node(data)
        
        if self.front is None:
            self.front = newNode
            self.back = self.front
        else:
            self.back.setNext(newNode)
            self.back = self.back.getNext()
        self.numlinks+=1
    
    def dequeue(self):
        if self.isEmpty():
            print("Empty Queue")
            return None
        frontNode = self.front
        if self.front.getNext() is None:
            self.front = None
            self.back = None
        else:
            self.front = self.front.getNext()
        self.numlinks-=1
        self.dqCount+=1
        return frontNode
    
    def qprint(self):
        if not self.isEmpty():
            frontNode = self.front
            while not frontNode is None:
                print(frontNode.getData(), end = " -> ")
                frontNode = frontNode.getNext()
            print("null")
        else:
            print("Empty Queue")
            
    def isEmpty(self):
        return self.numlinks == 0
    
    def size(self):
        return self.numlinks
    
class Stack:
    
    def __init__(self):
        self.head = None
        self.numlinks = 0
    
    def __str__(self):
        if self.isEmpty():
            print("Empty Stack")
        frontNode = self.head
        while frontNode is not None:
            print(frontNode.getData(), end = " -> ")
            frontNode = frontNode.getNext()
        print("null")
    
    def push(self, item):
        newNode = Node(item)
        newNode.next = self.head
        self.head = newNode
        self.numlinks+=1
    
    def pop(self):
        if self.isEmpty():
            print("Stack is Empty")
            return None
        topNode = self.head
        self.head = self.head.getNext()
        topNode.next = None
        self.numlinks-=1
        return topNode
    
    def isEmpty(self):
        return self.numlinks == 0
    
    def size(self):
        return self.numlinks

def main():
    q = Queue()
    n = 10000
    data = urllib.request.urlopen('https://raw.githubusercontent.com/gsprint23/cpts215/master/microassignments/files/words.txt').read()


    for i in range(len(data)):
        q.enqueue(data[i])
    start = time.time()
    while q.isEmpty() == False:
        q.dequeue()
    end = time.time()
    
    print(end - start)
    print('Dequeue Count', q.dqCount)
if __name__ == '__main__':
    main()
    

1.3950119018554688
Dequeue Count 1016714


#2/3 for Linked List

For a list of size 100

Time

1. 0.000499

2. 0.0

3. 0.0

4. 0.0

5. 0.0


Dequeue Count

1. 100

2. 100

3. 100

4. 100

5. 100

The linked list implementation is much faster than the list and two stacks implementations, with the dequeue function having an average runtime of $O(1)$ and worst case runtime of $O(n)$


For a list of size 1000

Time

1. 0.0010

2. 0.000996

3. 0.002

4. 0.001495

5. 0.0015


Dequeue Count

1. 1000

2. 1000

3. 1000

4. 1000

5. 1000

Still very fast, and with the same runtime. Next the function will be tested using a list size of 10000. For the two stacks implementation this took too long for the computer to handle, but the linked list implementation should be much faster.

For a list of size 1000

Time

1. 0.013993

2. 0.01401

3. 0.01299

4. 0.013009

5. 0.013028


Dequeue Count

1. 10000

2. 10000

3. 10000

4. 10000

5. 10000

Even for a list size of 10000, the linked list queue implementation is stil very fast. Thus, it is better to use the linked list implementation of the queue whenever possible.

Finally we can check using the words.txt file given. This was too much for both the list and two stacks implementation to handle.

Time

1. 1.3920485

2. 1.423613

3. 1.380976

4. 1.359086

5. 1.3950119


Dequeue Count

1. 1016714

2. 1016714

3. 1016714

4. 1016714

5. 1016714


Though the words.txt file has many words, the linked list implementation for queues still dequeues much faster when compared to the list and two stacks versions.

#Conclusion

Overall the linked list implementation of the queue is the fastest for the dequeue method. After this comes the list implementation and finally the two stacks implementation. The difference between these implementations is rather drastic. The two stacks implementation was too slow for this computer to handle for lists of size 10000. The list implementation was fine with a list size of 10000, but was too slow for the words.txt file. However the linked list implementation was extremely fast in all cases, even when the words.txt file was used. Thus, the linked list implementation of the queue should be used whenever possible as its dequeue function is extremely fast.